clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::higher::IfLetOrMatch;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::expr_visitor_no_bodies;
+use clippy_utils::{meets_msrv, msrvs, peel_blocks};
+use if_chain::if_chain;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::intravisit::Visitor;
+use rustc_hir::{Expr, ExprKind, MatchSource, Pat, PatKind, QPath, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Warn of cases where `let else` could be used
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `let else` provides a standard construct for this pattern
+    /// that people can easily recognize. It's also more compact.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # let w = Some(0);
+    /// let v = if let Some(v) = w { v } else { return };
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust
+    /// # #![feature(let_else)]
+    /// # fn main () {
+    /// # let w = Some(0);
+    /// let Some(v) = w else { return };
+    /// # }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub MANUAL_LET_ELSE,
+    restriction,
```

Comment:
Personally I think that `style` is where the lint should eventually end up, but there is currently no `rustfmt` support for `let else`. I'm not sure `style` should suggest something that is not well supported by the other tools. I think that users exist who might still want to be able to easily enable this lint in the meantime, so maybe a time inside `nursery` or `restriction` would be the best idea, and then once `rustfmt` formats it, it can be moved to `style`. What do you think?

Reply:
Good point about `rustfmt`. I would put it in `pedantic` then, because it will be tested there (from people that use the pedantic group). In restriction or nursery, only people that actively search for the lint will test it.

Reply:
I've put it into pedantic.